stream/src/main/java/org/apache/kafka/streaming/processor/internals/ProcessorContextImpl.java

```diff
@@ -0,0 +1,259 @@
+
+import org.apache.kafka.clients.consumer.Consumer;
+import org.apache.kafka.clients.consumer.KafkaConsumer;
+import org.apache.kafka.common.KafkaException;
+import org.apache.kafka.common.TopicPartition;
+import org.apache.kafka.common.metrics.Metrics;
+import org.apache.kafka.common.serialization.ByteArrayDeserializer;
+import org.apache.kafka.common.serialization.Deserializer;
+import org.apache.kafka.common.serialization.Serializer;
+import org.apache.kafka.streaming.StreamingConfig;
+import org.apache.kafka.streaming.processor.ProcessorContext;
+import org.apache.kafka.streaming.processor.StateStore;
+import org.apache.kafka.streaming.processor.Processor;
+import org.apache.kafka.streaming.processor.RestoreFunc;
+
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+public class ProcessorContextImpl implements ProcessorContext {
+
+    private static final Logger log = LoggerFactory.getLogger(ProcessorContextImpl.class);
+
+    private final int id;
+    private final StreamTask task;
+    private final Metrics metrics;
+    private final RecordCollector collector;
+    private final ProcessorStateManager stateMgr;
+    private final ArrayDeque<ProcessorNode> nodeStack = new ArrayDeque<ProcessorNode>();
+
+    private final Serializer<?> keySerializer;
+    private final Serializer<?> valSerializer;
+    private final Deserializer<?> keyDeserializer;
+    private final Deserializer<?> valDeserializer;
+
+    private boolean initialized;
+
+    @SuppressWarnings("unchecked")
+    public ProcessorContextImpl(int id,
+                                StreamTask task,
+                                StreamingConfig config,
+                                RecordCollector collector,
+                                Metrics metrics) throws IOException {
+        this.id = id;
+        this.task = task;
+        this.metrics = metrics;
+        this.collector = collector;
+
+        this.keySerializer = config.getConfiguredInstance(StreamingConfig.KEY_SERIALIZER_CLASS_CONFIG, Serializer.class);
+        this.valSerializer = config.getConfiguredInstance(StreamingConfig.VALUE_SERIALIZER_CLASS_CONFIG, Serializer.class);
+        this.keyDeserializer = config.getConfiguredInstance(StreamingConfig.KEY_DESERIALIZER_CLASS_CONFIG, Deserializer.class);
+        this.valDeserializer = config.getConfiguredInstance(StreamingConfig.VALUE_DESERIALIZER_CLASS_CONFIG, Deserializer.class);
+
+        File stateFile = new File(config.getString(StreamingConfig.STATE_DIR_CONFIG), Integer.toString(id));
+
+        Consumer restoreConsumer = new KafkaConsumer<>(
+            config.getConsumerConfigs(),
+            null /* no callback for restore consumer */,
+            new ByteArrayDeserializer(),
+            new ByteArrayDeserializer());
```

Comment:
Actually, I've written a `ProcessingTopologyTestDriver` class under `src/test` that takes a TopologyBuilder and will make it very easy for projects that use Kafka Streams to test their topologies with unit tests that do not use a real Kafka. Each test method can set up the driver, pass one or more methods to the driver (which then forwards them to the appropriate source), and finally check the messages output by the sinks. It's pretty simple, and it uses mock consumers and producers along with a single `StreamTask` to do all the heavy lifting. (This has the benefit of also testing the bulk of the `StreamTask` implementation.) 

So, it'd be great if this test driver could pass the consumer for state manager into the `StreamTask` constructor took a consumer (or consumer supplier) so that the tests can inject mocks instead; the `StreamTask` constructor already takes a consumer and producer, so taking a second consumer for state management seems consistent. Right now `StreamTask` is the only thing that constructs a `ProcessingContextImpl`, so passing the consumer down also seem reasonable. This approach also seems to have minimal impact on other code, and IMO is better than passing a flag into the processor state manager's constructor or calling a method on the processor state manager, since right now the the `ProcessorStateManager` is constructed within the `ProcessingContextImpl` class which itself is constructed within the `StreamTask` constructor.

BTW, the test driver class and a unit test is ready for a PR, except that the tests that use state management are failing right now because it's trying to create a real consumer for state management. I'll go ahead and make the aforementioned change to pass in the consumer, fix my tests, and submit a PR for review first thing tomorrow.